reps/2024-07-01-raycluster-crd-status-improvement.md

```diff
@@ -0,0 +1,137 @@
+
+2. Introduce `Status.Conditions` to surface errors with underlying Ray Pods. 
+    * TODO: We still need to define which types of information we want to surface in the CR status, operator logs, or Kubernetes events.
+
+3. Make suspending a RayCluster an atomic operation.
+    * Like RayJob, RayCluster also supports the [suspend](https://github.com/ray-project/kuberay/pull/1711) operation. However, the suspend operation is [atomic](https://github.com/ray-project/kuberay/pull/1798) in RayJob. In RayCluster, if we set suspend to true and then set it back to false before the RayCluster transitions to the state [rayv1.Suspended](https://github.com/ray-project/kuberay/blob/8453f9b6c749abf3b72b09e0d41032bbfcf17367/ray-operator/apis/ray/v1/raycluster_types.go#L115), there might be some side effects.
+
+4. Add a new condition `Status.Conditions[HeadReady]` to indicate whether the Ray head Pod is ready for requests or not. This is also useful for RayJob and RayService, where KubeRay needs to communicate with the Ray head Pod.
+
+5. Redefine `rayv1.Ready`.
+    * As mentioned in the “Context” section, when all Pods in the RayCluster are running for the first time, the RayCluster CR’s `Status.State` will transition to `ready` ([code](https://github.com/ray-project/kuberay/blob/8453f9b6c749abf3b72b09e0d41032bbfcf17367/ray-operator/apis/ray/v1/raycluster_types.go#L113)). It is highly possible that the RayCluster will remain in the `ready` state even if some incidents occur, such as manually deleting the head Pod.
+    * Here, we redefine `ready` as follows:
+        * When all Pods in the RayCluster are running for the first time, the RayCluster CR’s `Status.State` will transition to `ready`.
+        * After the RayCluster CR transitions to `ready`, KubeRay only checks whether the Ray head Pod is ready to determine if the RayCluster is `ready`.
+        * This new definition is mainly to maintain backward compatibility.
+    * TODO
+        * I am considering making `ready` a condition in `Status.Conditions` and removing `rayv1.Ready` from `Status.State`. Kubernetes Pod also makes a [PodCondition](https://pkg.go.dev/k8s.io/api/core/v1#PodConditionType), but this may be a bit aggressive.
+        * In the future, we can expose a way for users to self-define the definition of `ready` ([#1631](https://github.com/ray-project/kuberay/issues/1631)).
+
+## Implementation
+
+### Step 1: Introduce Status.Conditions
+
+```go
+type RayClusterStatus struct {
+       ...
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+type RayClusterConditionType string
+
+const (
+	HeadReady RayClusterConditionType = "HeadReady"
+	// Pod failures. See DeploymentReplicaFailure and
+// ReplicaSetReplicaFailure for more details.
+RayClusterReplicaFailure RayClusterConditionType = "ReplicaFailure"
```

Comment:
In a follow-up iteration of the conditions, I think a condition like `AllWorkersReady` indicating all worker replicas are ready will be useful too. The example that comes to mind is in RayJob controller to know when a job should start (assuming we deprecate the status.state field). Relying on head readiness won't be enough

Reply:
How about using `AllRayPodsReady`, which indicates that both the head and workers are ready? If we use `AllWorkersReady`, we need to check both `AllWorkersReady` and `HeadReady` to determine whether we can submit the job. If we have `AllRayPodsReady`, we only need to check `AllRayPodsReady`.